### PR TITLE
Mech Blue Updated & PIDF Wrapper

### PR DESCRIPTION
**Pull request by SiddharthKadari**
Added Wrapper class for CANSparkMax motors which includes PIDF in Velocity and Position calculations. PIDF constants must be tuned for each individual system, guide and supporter class for tuning for the purposes of our FRC robot yet to be made, this will be a future project.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -2,7 +2,7 @@ import edu.wpi.first.gradlerio.deploy.roborio.RoboRIO
 
 plugins {
     id "java"
-    id "edu.wpi.first.GradleRIO" version "2022.2.1"
+    id "edu.wpi.first.GradleRIO" version "2022.1.1"
```
Comment (AG6GR) [inline build.gradle]: We shouldn't be moving back versions. Can you check your local WPILib install is up to date?

File: src/main/java/frc/robot/systems/BallHandlingFSM.java [hunk 2]
```diff
@@ -9,25 +11,39 @@
...
-	private static final int PUSH_TIME_SECONDS = 3;
+	private static final double PUSH_TIME_SECONDS = 1;
+
+	private static final double INTAKE_MOTOR_RPM = 700;
 
 	/* ======================== Private variables ======================== */
 	private FSMState currentState;
 
 	private Solenoid pushSolenoid;
+	private Solenoid pullSolenoid;
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/BallHandlingFSM.java]: You probably are looking for DoubleSolenoid https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/DoubleSolenoid.html

File: src/main/java/frc/robot/wrappers/NeoSparkMaxPid.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+						CANSparkMax.MotorType.kBrushless);
+
+		motor.getPIDController().setP(p);
+		motor.getPIDController().setI(i);
+		motor.getPIDController().setD(d);
+		motor.getPIDController().setFF(f);
+		motor.getPIDController().setIZone(0);
+		motor.getPIDController().setOutputRange(-1, 1);
+
+		encoder = motor.getEncoder(SparkMaxRelativeEncoder.Type.kHallSensor, HALL_COUNTS_PER_REV);
```
Comment (AG6GR) [inline src/main/java/frc/robot/wrappers/NeoSparkMaxPid.java]: Does the basic getEncoder() not work here?

**Review by AG6GR — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/systems/BallHandlingFSM.java [hunk 4]
```diff
@@ -98,21 +135,78 @@ public void update(TeleopInput input) {
...
 	private FSMState nextState(TeleopInput input) {
+		if (input == null) {
+			return FSMState.IDLE;
+		}
+		System.out.println("Velocity: " + intakeMotor.getMotorTemperature());
+
 		switch (currentState) {
 			case IDLE:
-				if (input != null && input.isShooterButtonPressed()) {
+				if (pushCommandTime == -1 && input.isShooterButtonPressed()) {
```
Comment (tsoim) [inline src/main/java/frc/robot/systems/BallHandlingFSM.java]: please replace `-1` with a descriptive variable as it's not an obvious magic num

File: build.gradle (hunk 1 above)
Comment (SiddharthKadari) [reply build.gradle]: Resolved to 2022.3.1

File: src/main/java/frc/robot/systems/BallHandlingFSM.java (hunk 2 above)
Comment (SiddharthKadari) [reply src/main/java/frc/robot/systems/BallHandlingFSM.java]: Implemented

File: src/main/java/frc/robot/wrappers/NeoSparkMaxPid.java (hunk 3 above)
Comment (SiddharthKadari) [reply src/main/java/frc/robot/wrappers/NeoSparkMaxPid.java]: After closer inspection of the API, we believe it should, fixed

File: src/main/java/frc/robot/systems/BallHandlingFSM.java (hunk 4 above)
Comment (SiddharthKadari) [reply src/main/java/frc/robot/systems/BallHandlingFSM.java]: replaced with much better named variables

File: src/main/java/frc/robot/systems/BallHandlingFSM.java [hunk 5]
```diff
@@ -86,7 +89,7 @@ public FSMState getCurrentState() {
 	public void reset() {
 		currentState = FSMState.IDLE;
 
-		pushCommandTime = -1;
+		isSolenoidExtended = false;
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/BallHandlingFSM.java:92]: You can use pushSolenoid.get() to fetch the solenoid state, no need for separate state var. You can fix this in next patch set.

File: src/main/java/frc/robot/Robot.java [hunk 6]
```diff
@@ -7,6 +7,8 @@
 import com.revrobotics.REVPhysicsSim;
 
 import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.wpilibj.Compressor;
+import edu.wpi.first.wpilibj.PneumaticsModuleType;
 // WPILib Imports
```
Comment (AG6GR) [inline src/main/java/frc/robot/Robot.java:12]: Let's put the WPILib imports under the WPILib Imports section. You can fix this in the next patchset.

**Review by AG6GR — APPROVED** [review]

**Pull request merged**